### ISSUE TITLE
Priorities for Tridactyl retreat/getaway

### ISSUE THREAD

**Issue by bovine3dom** [tridactyl/tridactyl#1379]
A few of us Tridactyl developers are going to meet up in late May this year.

Let's use this issue to organise what we'll work on. All are welcome to contribute to this discussion and help us organise our priorities.

**Comment by glacambre** [comment]
I believe we really need to set up a testsuite.

**Comment by joaomsa** [comment]
UI responsiveness (specially for autocomplete)

**Comment by bovine3dom** [comment]
Somewhat related to UI responsiveness - we shouldn't ever have races on user input. #1208.

The random tabs #1237 stuff is really annoying, too. Especially when it eats my search terms.


It would be nice to resurrect the keyboard API, too.

**Comment by pmfaqtory** [comment]
Hi Guys,

Thank you very, very much for your tremendous work! I had thought that we would never see anything like Pentadactyl again, which would have been a tragedy. I was elated to find Tridactyl today and hope that you'll be able to make rapid progress.

I noticed that discussion about smooth scrolling dates back to nine months ago. Has there been any progress on that front? I urge you to make this your top priority, as it greatly affects usability. 

Have you considered a WordPress site and a manual? We could drive the number of users WAY up, once they realize what they're missing.

If I can be of help in any way, please let me know. Tridactyl is the best thing to ever happen to Firefox! Upon finding it, I immediately deleted SurfingKeys. :)

Thanks,

Steve

**Comment by bovine3dom** [comment]
None of the people who develop Tridactyl use smooth scrolling so it's not a priority for us at all (nor would we even know what good smooth scrolling was). I take it you have tried `:set smoothscroll true`? Could you explain what you dislike about it in another issue?

We briefly considered making a `github.io` page, but I personally think our GitHub readme is pretty well fine. None of us are web developers and we don't have much interest. We'd accept a PR that made a decent one - https://help.github.com/en/articles/user-organization-and-project-pages#project-pages-sites - gives you some instructions.

We've also considered opening up a GitHub wiki but no-one has really expressed any interest in it.

Are you aware of the manual available on `:tutor` and `:help`? We have considered making a standalone website out of them but, again, it's quite far down our list of priorities.

On all of the things you mentioned, we'd be happy to accept PRs and work with you to get them to a state where we can merge them.

I don't think any of us especially worry about the number of users we do or don't have - I'm much more concerned about how janky and bad Tridactyl is at the moment, because I notice that every day, whereas I notice the number of users about once a month or so whenever I can be bothered to check :)

**Comment by pmfaqtory** [comment]
Hi Oliver

I guess that it depends on what your goal is. If you’re developing Tridactyl for the use just of its developers, you should prioritize according to your needs. Personally, I wouldn’t switch from Chrome to Firefox for Tridactyl unless I had smooth scrolling. Not having it is a usability nightmare and in my opinion makes the browsing experience very shoddy, especially in 2019.

I’ve set smoothscroll to true, but it’s incredibly jerky. Not having buttery smooth scrolling is a show stopper for me, and will force me to continue using Chrome and SurfingKeys. Tridactyl is superior in every other way, though. I’d love to see it as a polished product with a broad audience, because it’s enormously useful. But I don’t think that you’ll see wide adoption without smooth scrolling working perfectly, because so many of us do care about it to the point of refusing to forego it under any circumstances. Using Tridactyl to scroll degrades the scrolling experience from what comes out-of-the-box, so to speak, with Firefox, and that really is a big problem.

Pentadactyl seemed to have many thousands of users. People knew about it, and it helped that it had its own website. Again, it really depends on what your goals are. I’d consider making a donation, but I’m leery about doing so for two reasons:

1. The Pentadactyl developers were hostile to Pentadactyl users, and apparently to one another. That never bodes well for the long-term survival of a product, and I’m not sure what to expect from Tridactyl (just to be honest); and

2. With such a small group of developers, it’s unclear whether Tridactyl might be abandoned at some point. No one would go all-in on Tridactyl if they knew that one year, or three years, from now, they’d lose something they’ve come to rely on. Sustainability is really important.

I think that both the user base and the development team need to reach critical mass. I don’t think that very many people know about Tridactyl, but that’s a solvable problem. Building a successful product involves a lot more than just software development. Marketing matters, and if you do it right, you could sell Tridactyl as a commercial product. I see many advantages to this, but it takes a lot of commitment and determination, with a high threshold for pain.

I hope that you’ll be able to make all of the improvements that you’re targeting, and I look forward to watching your progress. If you decide to pursue developing a commercial add-in, I would be interested in supporting you financially and in other ways. For now, though, I wish you good luck!

Best,

Steve

**Comment by raffomania** [comment]
Thanks so much for making tridactyl. It's been a huge step up from vimium.

I'd just like to leave my personal reason for using smooth scrolling here.
Smooth scrolling helps me keep my orientation when navigating a page. When jumping without a transition, I loose track of what I was looking at and need an extra 1 or 2 seconds to start reading again.

**Comment by bovine3dom** [comment]
A number of people in the wild have complained that Firefox steals focus from Tridactyl, and that they must therefore use Vimium / cVim etc in Chrome because of it.

It might be worth checking up on the 'competition' to see whether there are any good ideas we can borrow, or if there are any major issues that need fixing in Firefox that Chrome doesn't have.

**Comment by antonva** [comment]
We have a link to this [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1415860) in #760 and the issue is #1122

It seems to stem from out of process extensions for some reason.

**Comment by pirj** [comment]
Guys, thanks for this amazing extension!

I made several attempts in the past at using Vimium and Vimperator, and dropped it. It is completely different with Tridactyl, it went smooth, and I really enjoy it.

One thing that I'd love to see is fuzzy matching for `:o`. When I open a new tab, I, ergh, press `Tab :o`. I got used to fuzzy searchers for everything (using CtrlP), and it would really wipe out the differences in workflow to have it in Tridactyl.

Is it possible on opening a new tab to go command mode immediately by pressing `:`?

Wish you a fun and productive retreat!

**Comment by jordila** [comment]
Hi Trydactylers,
This extension gives me such a pleasent web browsing experience... ah.

Now that you ask,... i miss so much the support for Bookmarks tags #167  !

Me too, i wish you a fun and productive retreat !

**Comment by max-sixty** [comment]
A short message to say thank you! Tridactyl is the main reason I use Firefox vs Chrome. I made a donation as a token of appreciation. Keep up the great work!

**Comment by glittershark** [comment]
it seems small, but #64 is my biggest annoyance with tridactyl these days - to the point where I basically have to open Chrome to read things like dev documentation etc. that I'm doing a lot of finding in

**Comment by vtatai** [comment]
Just came here to leave my appreciation - Pentadactyl, and now Tridactyl, is the reason I've been able to hang on to Firefox after all these years. Also left a small contribution so you guys can keep on rocking!

**Comment by cmcaine** [comment]
> Guys, thanks for this amazing extension!
> 
> I made several attempts in the past at using Vimium and Vimperator, and dropped it. It is completely different with Tridactyl, it went smooth, and I really enjoy it.

Thanks for your comments @pirj!

> One thing that I'd love to see is fuzzy matching for `:o`. When I open a new tab, I, ergh, press `Tab :o`. I got used to fuzzy searchers for everything (using CtrlP), and it would really wipe out the differences in workflow to have it in Tridactyl.


We already have some fuzzy matching on `:o` and friends (we use fuse.js). Could you help me understand how you think it could be improved? Perhaps you could give an example of when an expected completion was not displayed?

We don't bind Ctrl-n or Ctrl-p at the moment, but we could add new binds for ctrl-p. Unfortunately, firefox won't let us take over the ctrl-n shortcut. At present we support tab and shift-tab for selecting completions.

> Is it possible on opening a new tab to go command mode immediately by pressing `:`?

If you open the new tab with ctrl-t, then firefox steals focus for the location bar and there's nothing we can do unless Moz let's us manage that behaviour :/

If you open with `:tabopen` (bound to t), then you should be able to open command mode pretty quickly by pressing `:`. Similarly if you open a new tab from a link with hint mode or the mouse.

At the moment, we have to wait for the page to load at least a bit, is this the delay you're thinking about?

**Comment by pirj** [comment]
> an example of when an expected completion was not displayed?
Sure!

![image](https://user-images.githubusercontent.com/6916/57927877-04b81800-78b8-11e9-9a34-add5b5582503.png)

I would expect some links to appear for `rubr` (first five match, seventh from the top, and the one in the bottom as well), but none are matched:
![image](https://user-images.githubusercontent.com/6916/57927959-2c0ee500-78b8-11e9-90b0-7d9cd6b49494.png)

> we support tab and shift-tab for selecting completions.
That's perfectly fine.

> If you open with :tabopen
Awesome, that's what I was looking for. 👍 Breaking my habit of pressing `C-t` now.

> page to load at least a bit, is this the delay you're thinking about?
No, there are no delays, I meant I have to press Tab first to remove focus from the location bar. `:tabopen` is just what I was looking for. 👍

**Comment by dit7ya** [comment]
One universal text based config dotfile for everything.

**Comment by glacambre** [comment]
@dit7ya Can you explain in detail what the differences between what you're suggesting and the current `.tridactylrc` implementation are?

**Comment by bovine3dom** [comment]
Probably worth harvesting most-upvoted issues from here: https://github.com/tridactyl/tridactyl/issues?utf8=%E2%9C%93&q=+sort%3Areactions

**Comment by bovine3dom** [comment]
There's a short [dev diary on the new tab page](https://github.com/tridactyl/tridactyl/blob/050bc66364cd1d047d458875a9aabe170fb753c9/src/static/newtab.md#user-content-tridactyl-replace_me_with_the_version_using_sed) and a [changelog for 1.16.0](https://github.com/tridactyl/tridactyl/blob/7312a0868ffbe2f555cd39250c88012cabfc5d14/CHANGELOG.md#user-content-release-1160--unreleased). I made a [retreats](https://github.com/tridactyl/retreats) repository which may turn into a blog post, time permitting. It currently just houses the code for generating a Sankey diagram for our expenses.

In total, we closed about 60 issues and opened about 30 new ones, with around 110 commits to master.

With respect to issues raised in this thread:

- [x] Test suite (slightly temperamental, but it exists: #1556
- [x] UI responsiveness (OK, cmcaine did this this morning, but I'm still going to count it: #1655 
- [ ] Random tabs - not done yet. We've made progress towards rewriting Tridactyl's state in a more functional manner which should help, but it's not merged (or even a PR) yet. Watch [this space](https://github.com/cmcaine/tridactyl-2-experiments).
- [ ] (Our) keyboard-api: We made some progress towards getting the experiment working again, but it's nothing to write home about: see [this repo](https://github.com/tridactyl/keyboard-api)
- [ ] Smooth scrolling:
    - [x] Partially fixed in #1623 for single key presses
    - [ ] Blocked by #1645 for continuous 'good' smooth scrolling (with the caveat that all of the core developers believe that this is a contradiction in terms)
- [ ] Website: I still don't think Tridactyl is actually good enough to actively advertise. Perhaps once we have the keyboard-api and random tabs fixed.
- [ ] Commercial add-on: I opened #1619, but it's mostly tongue-in-cheek.
- [x] Firefox stealing focus from Tridactyl: I'm wondering whether this was actually about our error messages stealing focus, which I did fix.
- [ ] Fuzzy matching for completions: We'll try to look into this once we switch completions to Mithril.
- [ ] Bookmarks tags (#167): idem
- [x] Find again (#64): done - albeit without `incsearch`. Still left unbound by default.
- [x] One universal text based config dotfile for everything: I'm going to take the liberty of assuming this meant `mkt`, #695.


Thanks to all of our donors who helped to make the week possible. :heart:
